upup/pkg/fi/cloudup/apply_cluster.go

```diff
@@ -1422,6 +1424,41 @@ func (n *nodeUpConfigBuilder) BuildConfig(ig *kops.InstanceGroup, apiserverAddit
 		}
 	}
 
+	if cluster.Spec.GetCloudProvider() == kops.CloudProviderOpenstack {
+		keyset := keysets["kops-controller-bootstrap"]
+		if !ig.HasAPIServer() {
+			if keyset != nil && keyset.Primary != nil && keyset.Primary.PrivateKey != nil {
```

Comment:
If the keyset isn't there or doesn't have a primary private key, it should error out.